core/src/main/scala/org/apache/spark/util/collection/TieredDiskMerger.scala

```diff
@@ -0,0 +1,232 @@
+    val mergedItr =
+      MergeUtil.mergeSort(finalItrGroup, keyComparator, dep.keyOrdering, dep.aggregator)
+
+    onDiskBlocks.clear()
+
+    // Release the on-disk file when iteration is completed.
+    val completionItr = CompletionIterator[Product2[K, C], Iterator[Product2[K, C]]](
+      mergedItr, releaseShuffleBlocks(finalMergedBlocks))
+
+    new InterruptibleIterator(context, completionItr)
+  }
+
+  def start() {
+    diskToDiskMerger.start()
+  }
+
+  /**
+   * Release the left in-memory buffer or on-disk file after merged.
+   */
+  private def releaseShuffleBlocks(onDiskShuffleGroup: Array[DiskShuffleBlock]): Unit = {
+    onDiskShuffleGroup.map { case DiskShuffleBlock(_, file, _) =>
+      try {
+        logDebug(s"Deleting the unused temp shuffle file: ${file.getName}")
+        file.delete()
+      } catch {
+        // Swallow the exception
+        case e: Exception => logWarning(s"Unexpected errors when deleting file: ${
+          file.getAbsolutePath}", e)
+      }
+    }
+  }
+
+  private def onDiskBlocksToIterators(shufflePartGroup: Seq[DiskShuffleBlock])
+    : Seq[Iterator[Product2[K, C]]] = {
+    shufflePartGroup.map { case DiskShuffleBlock(id, _, _) =>
+      blockManager.diskStore.getValues(id, ser).get.asInstanceOf[Iterator[Product2[K, C]]]
+    }.toSeq
+  }
+
+  /**
+   * Whether we should carry out a disk-to-disk merge now or wait for more blocks or a done
+   * registering notification to come in.
+   *
+   * We want to avoid merging more blocks than we need to. Our last disk-to-disk merge may
+   * merge fewer than maxMergeFactor blocks, as its only requirement is that, after it has been
+   * carried out, <= maxMergeFactor blocks remain. E.g., if maxMergeFactor is 10, no more blocks
+   * will come in, and we have 13 on-disk blocks, the optimal number of blocks to include in the
+   * last disk-to-disk merge is 4.
+   *
+   * While blocks are still coming in, we don't know the optimal number, so we hold off until we
+   * either receive the notification that no more blocks are coming in, or until maxMergeFactor
+   * merge is required no matter what.
+   *
+   * E.g. if maxMergeFactor is 10 and we have 19 or more on-disk blocks, a 10-block merge will put
+   * us at 10 or more blocks, so we might as well carry it out now.
+   */
+  private def shouldMergeNow(): Boolean = doneRegistering ||
+    onDiskBlocks.size() >= maxMergeFactor * 2 - 1
+
+  private final class DiskToDiskMerger extends Thread {
+    setName(s"tiered-merge-thread-${Thread.currentThread().getId}")
+    setDaemon(true)
+
+    override def run() {
+      // Each iteration of this loop carries out a disk-to-disk merge. We remain in this loop until
+      // no more disk-to-disk merges need to be carried out, i.e. when no more blocks are coming in
+      // and the final merge won't need to merge more than maxMergeFactor blocks.
+      while (!doneRegistering || onDiskBlocks.size() > maxMergeFactor) {
+        while (!shouldMergeNow()) {
```

Comment:
here, i think we can rewrite the following:
mergeReadyMonitor.synchronized {
    if (!shouldMergeNow()) {
        mergeReadyMonitor.wait()
   }
}